packages/mui-base/src/useSwitch/useSwitch.ts

```diff
@@ -51,79 +51,46 @@ export function useSwitch(props: UseSwitchParameters): UseSwitchReturnValue {
       otherProps.onChange?.(event);
     };
 
-  const {
-    isFocusVisibleRef,
-    onBlur: handleBlurVisible,
-    onFocus: handleFocusVisible,
-    ref: focusVisibleRef,
-  } = useIsFocusVisible();
-
-  const [focusVisible, setFocusVisible] = React.useState(false);
-  if (disabled && focusVisible) {
-    setFocusVisible(false);
-  }
-
-  React.useEffect(() => {
-    isFocusVisibleRef.current = focusVisible;
-  }, [focusVisible, isFocusVisibleRef]);
-
-  const inputRef = React.useRef<HTMLInputElement | null>(null);
-
-  const createHandleFocus =
-    (otherProps: React.InputHTMLAttributes<HTMLInputElement>) =>
-    (event: React.FocusEvent<HTMLInputElement>) => {
-      // Fix for https://github.com/facebook/react/issues/7769
-      if (!inputRef.current) {
-        inputRef.current = event.currentTarget;
-      }
-
-      handleFocusVisible(event);
-      if (isFocusVisibleRef.current === true) {
-        setFocusVisible(true);
-        onFocusVisible?.(event);
-      }
-
-      onFocus?.(event);
-      otherProps.onFocus?.(event);
-    };
-
-  const createHandleBlur =
-    (otherProps: React.InputHTMLAttributes<HTMLInputElement>) =>
-    (event: React.FocusEvent<HTMLInputElement>) => {
-      handleBlurVisible(event);
-
-      if (isFocusVisibleRef.current === false) {
-        setFocusVisible(false);
+  const createHandleClick =
+    (otherProps: React.ButtonHTMLAttributes<HTMLButtonElement>) =>
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      otherProps.onClick?.(event);
+      if (event.defaultPrevented || readOnly) {
+        return;
       }
 
-      onBlur?.(event);
-      otherProps.onBlur?.(event);
+      inputRef.current?.click();
     };
 
-  const handleInputRef = useForkRef(focusVisibleRef, inputRef);
+  const getButtonProps: UseSwitchReturnValue['getButtonProps'] = (otherProps = {}) => ({
+    ...otherProps,
+    type: 'button',
+    role: 'switch',
+    'aria-checked': checked,
+    'aria-disabled': disabled,
+    'aria-readonly': readOnly,
+    onClick: createHandleClick(otherProps),
+  });
```

Comment:
The prop getters here should likely be memoized, so they can be passed to a memo'd component via the `render` prop where necessary. I also think it's cleaner to inline the event handlers inside the props themselves?

Reply:
Done. As for inlining - I'm not entirely convinced. I don't like too many levels of indentation. I moved the event handlers closer to these functions so it's easier to see both at the same time

Reply:
I'm mostly fine with either, but I do think inlining is simpler and the better option:

- You don't need to specify types for the parameters, and there's less indirection
- It's easier to read compared to the double function
- It's colocated with the other props
- The indentation is the same (8 spaces):

### Separate

```js
  const createHandleButtonClick = React.useCallback(
    (otherProps: React.ButtonHTMLAttributes<HTMLButtonElement>) =>
      (event: React.MouseEvent<HTMLButtonElement>) => {
        inputRef.current?.click();
12345678^
      },
    [readOnly],
  );

  const getButtonProps: UseSwitchReturnValue['getButtonProps'] = React.useCallback(
    (otherProps = {}) => ({
      onClick: createHandleButtonClick(otherProps),
    }),
    [checked, disabled, readOnly, createHandleButtonClick],
  );
```

### Inline

```js
  const getButtonProps: UseSwitchReturnValue['getButtonProps'] = React.useCallback(
    (otherProps = {}) => ({
      onClick(event) {
        inputRef.current?.click();
12345678^
      },
    }),
    [checked, disabled, readOnly],
  );
```

Overall, I'm not that concerned and think either is fine, but I feel inlining is easier to read and write.